### ISSUE TITLE
Unexpected number of iterations with multiprocess minibatch loading

### ISSUE THREAD

**Issue by yoshum** [facebookresearch/pytorchvideo#102]
## 🐛 Bugs / Unexpected behaviors

If a dataset prepared as a `LabeledVideoDataset` instance is loaded with `num_workers > 1` and `batch_size > 1`, the number of batches loaded is different from what is expected, which is `num_videos // batch_size` with a single-view, single-clip-per-video dataset and `drop_last=True`. With `drop_last=False`, `DataLoader` yields unexpectedly many batches with the sizes less than `batch_size`. Apparently, the number of iterations is fixed to a multiple of `num_workers`.

This is generally not very crutial; it just results in the slightly smaller number of training iterations per epoch and the slightly more eval iterations. But it can cause some trouble when the number of iterations needs to be guessed.

## Instructions To Reproduce the Issue:

Code to reproduce the issue:
```
import os

from pytorchvideo.data import LabeledVideoDataset
from pytorchvideo.data.clip_sampling import RandomClipSampler
from pytorchvideo.transforms import ApplyTransformToKey, UniformTemporalSubsample
from torch.utils.data import DataLoader
from torchvision.transforms import Compose, RandomCrop

num_videos = 24

# generate labeled_video_path with num_videos==24.
data_dir = "/path/to/dir/with/many/videos"
files = os.listdir(data_dir)
labeled_video_path = [
    (os.path.join(data_dir, files[i]), {"label": 0}) for i in range(num_videos)
]

transform = ApplyTransformToKey(
    key="video", transform=Compose([UniformTemporalSubsample(8), RandomCrop(32)])
)
dataset = LabeledVideoDataset(
    labeled_video_path,
    clip_sampler=RandomClipSampler(2),
    transform=transform,
    decode_audio=False,
)

for num_workers in (8, 1):
    for batch_size in (2, 1):
        for drop_last in (True, False):
            loader = DataLoader(
                dataset,
                batch_size=batch_size,
                num_workers=num_workers,
                drop_last=drop_last,
            )
            num_iters = len([batch for batch in loader])
            print(
                f"ACTUAL: {num_iters}, EXPECTED: {num_videos // batch_size} "
                f"(batch_size={batch_size}, num_workers={num_workers}, drop_last={drop_last})"
            )
```

Output:
```
ACTUAL: 8, EXPECTED: 12 (batch_size=2, num_workers=8, drop_last=True)
ACTUAL: 16, EXPECTED: 12 (batch_size=2, num_workers=8, drop_last=False)
ACTUAL: 24, EXPECTED: 24 (batch_size=1, num_workers=8, drop_last=True)
ACTUAL: 24, EXPECTED: 24 (batch_size=1, num_workers=8, drop_last=False)
ACTUAL: 12, EXPECTED: 12 (batch_size=2, num_workers=1, drop_last=True)
ACTUAL: 12, EXPECTED: 12 (batch_size=2, num_workers=1, drop_last=False)
ACTUAL: 24, EXPECTED: 24 (batch_size=1, num_workers=1, drop_last=True)
ACTUAL: 24, EXPECTED: 24 (batch_size=1, num_workers=1, drop_last=False)
```

**Comment by juliendenize** [comment]
I think it is expected as the LabeledVideoDataset inherits IterableDataset from Pytorch.

In [Pytorch documentation](https://pytorch.org/docs/stable/data.html#loading-batched-and-non-batched-data) on data I found this note:
> When fetching from [iterable-style datasets](https://pytorch.org/docs/stable/data.html#iterable-style-datasets) with [multi-processing](https://pytorch.org/docs/stable/data.html#multi-process-data-loading), the drop_last argument drops the last non-full batch of each worker’s dataset replica.

I think that in order to have the right amount of batches, the labeled video datasets should follow map-style datasets. I understand why the iterable-style dataset is interesting especially when the stream of data is untrustworthy but it also has its downsides.

**Comment by yoshum** [comment]
@juliendenize No, it's not a problem inherent in IterableDataset, but in how LabeledVideoDataset is implemented. Note the following warning on [the same page](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader):

> `len(dataloader)` heuristic is based on the length of the sampler used. When dataset is an [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset), it instead returns an estimate based on `len(dataset) / batch_size`, with proper rounding depending on `drop_last`, regardless of multi-process loading configurations. This represents the best guess PyTorch can make because PyTorch trusts user dataset code in correctly handling multi-process loading to avoid duplicate data.
> However, if sharding results in multiple workers having incomplete last batches, this estimate can still be inaccurate, because (1) an otherwise complete batch can be broken into multiple ones and (2) more than one batch worth of samples can be dropped when `drop_last` is set. Unfortunately, PyTorch can not detect such cases in general.

I think the second paragraph describes exactly what is happening now. Current implementation of `LabeledVideoDataset` distributes data evenly to the replicas, resulting in multiple workers having incomplete last batches. One can alternatively implement `LabeledVideoDataset` so that it assign to all but at most one replicas data whose length is a multiple of the batch size. Then it should behave as expected. In this case, some workers might produce one more batch than the others, but that's OK.

I would say that `len(dataloader)` is always expected to be `len(dataset) / batch_size` with proper rounding. It is just an expectation and not a requirement, but fulfilling it is desirable when it's possible.

**Comment by juliendenize** [comment]
Yes, you are right, I'm currently trying to implement a `LabeledVideoDataset` in my project to alleviate this issue, but I didn't think of a proper way to pass the batch size to the `MultiProcessSampler` as it is called by the dataset and I don't like the idea to pass the `batch_size` to the dataset.

I think I will follow PySlowFast repo [Kinetics implementation](https://github.com/facebookresearch/SlowFast/blob/main/slowfast/datasets/kinetics.py) and implement a map-style dataset. As far as I understand it, it doesn't require a lot of changes in `LabeledVideoDataset` and it's nice to have `__getitem__` and `__len__` implemented. It should also remove the need to specify how the workers operate. Also, it would be easier to use pytorch-lightning without requiring to specify the current epoch for fixing seed at each epoch for example. I don't know the downsides of using this kind of datasets in comparison with Iterable though.
